Sources/PackageSyntax/PackageEditor.swift

```diff
@@ -0,0 +1,469 @@
+    case executable(name: String, dependencyNames: [String])
+    case test(name: String, dependencyNames: [String])
+    case binary(name: String, urlOrPath: String, checksum: String?)
+
+    /// The name of the factory method for a target type.
+    func factoryMethodName(for toolsVersion: ToolsVersion) -> String {
+        switch self {
+        case .executable:
+          if toolsVersion >= .v5_4 {
+            return "executableTarget"
+          } else {
+            return "target"
+          }
+        case .library: return "target"
+        case .test: return "testTarget"
+        case .binary: return "binaryTarget"
+        }
+    }
+
+    /// The name of the new target.
+    var name: String {
+        switch self {
+        case .library(name: let name, includeTestTarget: _, dependencyNames: _),
+             .executable(name: let name, dependencyNames: _),
+             .test(name: let name, dependencyNames: _),
+             .binary(name: let name, urlOrPath: _, checksum: _):
+            return name
+        }
+    }
+}
+
+public enum PackageDependencyRequirement: Equatable {
+    case exact(String)
+    case revision(String)
+    case branch(String)
+    case upToNextMajor(String)
+    case upToNextMinor(String)
+    case range(String, String)
+    case closedRange(String, String)
+    case localPackage
+
+    var ref: String? {
+        switch self {
+        case .exact(let ref): return ref
+        case .revision(let ref): return ref
+        case .branch(let ref): return ref
+        case .upToNextMajor(let ref): return ref
+        case .upToNextMinor(let ref): return ref
+        case .range(let start, _): return start
+        case .closedRange(let start, _): return start
+        case .localPackage: return nil
+        }
+    }
+}
+
+extension ProductType {
+    var isLibrary: Bool {
+        switch self {
+        case .library:
+            return true
+        case .executable, .test, .plugin, .snippet:
+            return false
+        }
+    }
+}
+
+/// The global context for package editor.
+public final class PackageEditorContext {
+    /// Path to the package manifest.
+    let manifestPath: AbsolutePath
+
+    /// The manifest loader.
+    let manifestLoader: ManifestLoaderProtocol
+
+    /// The repository manager.
+    let repositoryManager: RepositoryManager
+
+    /// The file system in use.
+    let fs: FileSystem
+
+    /// The diagnostics engine used to report errors.
+    let diagnosticsEngine: DiagnosticsEngine
+
+    public init(manifestPath: AbsolutePath,
+                repositoryManager: RepositoryManager,
+                toolchain: UserToolchain,
+                diagnosticsEngine: DiagnosticsEngine,
+                fs: FileSystem = localFileSystem) throws {
+        self.manifestPath = manifestPath
+        self.repositoryManager = repositoryManager
+        self.diagnosticsEngine = diagnosticsEngine
+        self.fs = fs
+
+        self.manifestLoader = ManifestLoader(toolchain: toolchain)
```

Comment:
probably use the same one we use for the workspace. see comment above about improving this initializer